README.md
# SN Update Set Extractor

A ServiceNow utility designed to streamline the process of extracting code from update sets.

## Overview

This utility, written in Go, processes ServiceNow update sets. It extracts and organizes code from XML files into corresponding directories and files, making it easier to use in an IDE.

## Prerequisites

- [Go (1.16 or later)](https://go.dev/doc/install)

## Getting Started

Clone the repository & change directory:

```sh
git clone https://github.com/stevengregory/sn-update-set-extractor
cd sn-update-set-extractor
```

## Load Update Sets

In the root of the project, create a `data` directory.

```sh
mkdir data
```

Load your update sets into this directory. When running the project, the utility will process all XML files it contains.

## Building & Running

To build and run the project, use the following commands:

```sh
make clean  # Remove any previous build and output files
make build  # Compile the Go project
make run    # Execute the compiled binary
```

Run the `make` command to perform all of these steps in one go:

```sh
make
```

## Generated Output

After building & running the project, the extracted code is structured and saved in the `dist` directory. Each file type is categorized under a parent directory for easy navigation. For example, the output directory structure might look like:

```
dist/
├── Client Development/
│   ├── Client Script/
│   └── UI Script/
├── Server Development/
│   ├── Business Rule/
│   ├── Fix Script/
│   ├── Script Include/
│   └── UI Action/
├── Service Portal/
│   ├── Angular ng-template/
│   ├── Header | Footer/
│   ├── Page/
│   ├── Style Sheet/
│   ├── Theme/
│   └── Widget/
│   └── Widget Angular Provider/
└── Inbound Integrations/
    └── Scripted REST Resource/
```

## Supported File Types

Currently, the extraction & organization of the following ServiceNow file types is supported:

- Angular ng-template
- Business Rule
- Client Script
- Fix Script
- Header | Footer
- Page
- Script Include
- Scripted REST Resource
- Style Sheet
- Theme
- UI Action
- UI Script
- Widget
- Widget Angular Provider

Efforts to support additional file types are ongoing. If there's a specific file type you'd like to see supported, please create an issue in the GitHub repository.

## Changelog

For a detailed breakdown of all changes made to this project, see the [CHANGELOG](CHANGELOG.md) file.

## License

This project is licensed under the terms of the MIT license. See the [LICENSE](LICENSE) file.


CHANGELOG.md
# Changelog

## [1.6.1](https://github.com/stevengregory/sn-update-set-extractor/releases/tag/v1.6.1) (2023-08-05)

### Fixed

- Corrected the `FixScript` struct in the XML parser to properly extract the `script` field from XML files. Previously, the parser was set to extract the `description` field.

## [1.6.0](https://github.com/stevengregory/sn-update-set-extractor/releases/tag/v1.6.0) (2023-07-25)

### Added

- Support for the "Widget Angular Provider" file type. The code extractor can now handle files associated with "Widget Angular Provider" types.
- Updated the file operations and helper functions to include handling for "Widget Angular Provider" files.
- README and CHANGELOG updates to include the information about "Widget Angular Provider" support.

## [1.5.0](https://github.com/stevengregory/sn-update-set-extractor/releases/tag/v1.5.0) (2023-06-20)

### Added

- Support for the "Style Sheet" file type. The code extractor can now handle SCSS files associated with "Style Sheet" types.
- README and CHANGELOG updates to include the information about Style Sheet support.

## [1.4.0](https://github.com/stevengregory/sn-update-set-extractor/releases/tag/v1.4.0) (2023-06-19)

### Added

- Support for the "Page" file type. The code extractor can now handle Page configurations associated with "Page" files.
- README and CHANGELOG updates to include the information about Page support.
- Ignoring `/tmp` and `/temp` directories in the `.gitignore` to avoid committing temporary files.

### Refactored

- Removed the unused `supportedFileTypes` function for cleaner codebase.

## [1.3.0](https://github.com/stevengregory/sn-update-set-extractor/releases/tag/v1.3.0) (2023-06-18)

### Added

- Reorganized the file directory structure for better navigation and categorization of scripts.
- Update the README with the new directory structure.

## [1.2.0](https://github.com/stevengregory/sn-update-set-extractor/releases/tag/v1.2.0) (2023-06-17)

### Added

- Support for the "Angular ng-template" file type. The code extractor can now handle HTML templates associated with "Angular ng-template" files.
- README and CHANGELOG updates to include the information about Angular ng-template support.

## [1.1.0](https://github.com/stevengregory/sn-update-set-extractor/releases/tag/v1.1.0) (2023-06-16)

### Added

- Support for the "Theme" file type. The code extractor can now handle SASS stylesheets associated with "Theme" files.
- README and CHANGELOG updates to include the information about UI Action support.

## [1.0.0](https://github.com/stevengregory/sn-update-set-extractor/releases/tag/v1.0.0) (2023-06-15)

### Added

- Initial public release, providing a utility to streamline the process of extracting code from ServiceNow update sets.
- Automatically parses XML files from the `data` directory, and organizes the extracted code into corresponding directories and files in the `dist` directory.
- Supported file types: Business Rule, Client Script, Fix Script, Header | Footer, Script Include, Scripted REST Resource, UI Script, Widget, and UI Action.
- Includes a Makefile for easy build and execution.
- Provides a detailed README.md for usage and project understanding.


xmlparser.go
package main

import (
	"encoding/xml"
	"fmt"
	"io/ioutil"
	"os"
	"path/filepath"
	"strings"
)

type Unload struct {
	XMLName    xml.Name    `xml:"unload"`
	UpdateSets []UpdateSet `xml:"sys_remote_update_set"`
	XMLScripts []XMLScript `xml:"sys_update_xml"`
}

type UpdateSet struct {
	Action          string `xml:"action,attr"`
	Application     string `xml:"application"`
	ApplicationName string `xml:"application_name"`
}

type XMLScript struct {
	Action      string `xml:"action"`
	Application string `xml:"application"`
	Name        string `xml:"name"`
	Type        string `xml:"type"`
	Payload     string `xml:"payload"`
}

type Widget struct {
	XMLName      xml.Name `xml:"sp_widget"`
	ClientScript string   `xml:"client_script"`
	Css          string   `xml:"css"`
	Script       string   `xml:"script"`
	Template     string   `xml:"template"`
	OptionSchema string   `xml:"option_schema"`
	Link         string   `xml:"link"`
}

type RecordUpdate struct {
	XMLName xml.Name `xml:"record_update"`
	Widget  Widget   `xml:"sp_widget"`
}

func parseXMLFile(filePath string) (*Unload, error) {
	xmlData, err := ioutil.ReadFile(filePath)
	if err != nil {
		return nil, err
	}

	var unload Unload
	err = xml.Unmarshal(xmlData, &unload)
	if err != nil {
		return nil, err
	}

	return &unload, nil
}

func extractCDATA(content string) string {
	jsContent := strings.TrimSpace(content)

	jsCodeStart := strings.Index(jsContent, "<![CDATA[")
	jsCodeEnd := strings.LastIndex(jsContent, "]]>")
	if jsCodeStart != -1 && jsCodeEnd != -1 {
		jsContent = jsContent[jsCodeStart+len("<![CDATA[") : jsCodeEnd]
	}

	return jsContent
}

func createDirectoryStructureAndFiles(unload *Unload, outputDir string) error {
	widgetFileTypes := map[string]string{
		"client_script": "client-script.js",
		"css":           "style.scss",
		"script":        "script.js",
		"template":      "template.html",
		"option_schema": "options.json",
		"link":          "link.js",
	}

	for _, script := range unload.XMLScripts {
		if script.Type == "System Property" {
			continue
		}

		dirPath := filepath.Join(outputDir, script.Type)
		if err := os.MkdirAll(dirPath, 0755); err != nil {
			return err
		}

		if script.Type == "Widget" {
			var recordUpdate RecordUpdate
			err := xml.Unmarshal([]byte(script.Payload), &recordUpdate)
			if err != nil {
				fmt.Printf("Failed to parse widget: %v\n", err)
				continue
			}

			// Create directory for the widget
			widgetDirPath := filepath.Join(dirPath, script.Name)
			if err := os.MkdirAll(widgetDirPath, 0755); err != nil {
				return err
			}

			widget := recordUpdate.Widget
			jsContent := map[string]string{
				"client_script": extractCDATA(widget.ClientScript),
				"css":           extractCDATA(widget.Css),
				"script":        extractCDATA(widget.Script),
				"template":      extractCDATA(widget.Template),
				"option_schema": extractCDATA(widget.OptionSchema),
				"link":          extractCDATA(widget.Link),
			}

			for key, value := range jsContent {
				fileName := widgetFileTypes[key]
				filePath := filepath.Join(widgetDirPath, fileName)
				if err := ioutil.WriteFile(filePath, []byte(value), 0644); err != nil {
					return err
				}
			}

		} else {
			fileName := fmt.Sprintf("%s.js", strings.ToLower(script.Name))
			filePath := filepath.Join(dirPath, fileName)
			jsContent := extractCDATA(script.Payload)
			if err := ioutil.WriteFile(filePath, []byte(jsContent), 0644); err != nil {
				return err
			}
		}
	}

	return nil
}

func main() {
	xmlFilePath := os.Args[1]
	outputDir := os.Args[2]

	unload, err := parseXMLFile(xmlFilePath)
	if err != nil {
		fmt.Println("Error parsing XML:", err)
		os.Exit(1)
	}

	err = createDirectoryStructureAndFiles(unload, outputDir)
	if err != nil {
		fmt.Println("Error creating directory structure and files:", err)
		os.Exit(1)
	}

	fmt.Println("Directories and files successfully created in", outputDir)
}


internal/xmlparser/types.go
package xmlparser

import "encoding/xml"

type Unload struct {
	XMLName    xml.Name    `xml:"unload"`
	UpdateSets []UpdateSet `xml:"sys_remote_update_set"`
	XMLScripts []XMLScript `xml:"sys_update_xml"`
}

type UpdateSet struct {
	Action          string `xml:"action,attr"`
	Application     string `xml:"application"`
	ApplicationName string `xml:"application_name"`
}

type XMLScript struct {
	Action      string `xml:"action"`
	Application string `xml:"application"`
	Name        string `xml:"name"`
	Type        string `xml:"type"`
	Payload     string `xml:"payload"`
	TargetName  string `xml:"target_name"`
}

type Widget struct {
	XMLName      xml.Name `xml:"sp_widget"`
	ClientScript string   `xml:"client_script"`
	Css          string   `xml:"css"`
	Script       string   `xml:"script"`
	Template     string   `xml:"template"`
	OptionSchema string   `xml:"option_schema"`
	Link         string   `xml:"link"`
}

type HeaderFooter struct {
	XMLName      xml.Name `xml:"sp_header_footer"`
	ClientScript string   `xml:"client_script"`
	Css          string   `xml:"css"`
	Script       string   `xml:"script"`
	Template     string   `xml:"template"`
	OptionSchema string   `xml:"option_schema"`
	Link         string   `xml:"link"`
}

type FixScript struct {
	Script string `xml:"script"`
}

type Page struct {
	Css string `xml:"css"`
}

type StyleSheet struct {
	Css string `xml:"css"`
}

type Theme struct {
	CssVariables string `xml:"css_variables"`
}

type UIAction struct {
	Script string `xml:"script"`
}

type RecordUpdate struct {
	XMLName      xml.Name     `xml:"record_update"`
	TargetName   string       `xml:"target_name"`
	Widget       Widget       `xml:"sp_widget"`
	HeaderFooter HeaderFooter `xml:"sp_header_footer"`
	FixScript    FixScript    `xml:"sys_script_fix"`
	UIAction     UIAction     `xml:"sys_ui_action"`
	Theme        Theme        `xml:"sp_theme"`
	Page         Page         `xml:"sp_page"`
	StyleSheet   StyleSheet   `xml:"sp_css"`
}

type WidgetContent interface {
	GetClientScript() string
	GetCss() string
	GetScript() string
	GetTemplate() string
	GetOptionSchema() string
	GetLink() string
}

func (w Widget) GetClientScript() string { return w.ClientScript }
func (w Widget) GetCss() string          { return w.Css }
func (w Widget) GetScript() string       { return w.Script }
func (w Widget) GetTemplate() string     { return w.Template }
func (w Widget) GetOptionSchema() string { return w.OptionSchema }
func (w Widget) GetLink() string         { return w.Link }

func (h HeaderFooter) GetClientScript() string { return h.ClientScript }
func (h HeaderFooter) GetCss() string          { return h.Css }
func (h HeaderFooter) GetScript() string       { return h.Script }
func (h HeaderFooter) GetTemplate() string     { return h.Template }
func (h HeaderFooter) GetOptionSchema() string { return h.OptionSchema }
func (h HeaderFooter) GetLink() string         { return h.Link }


cmd/sn-update-set-extractor/main.go
package main

import (
	"fmt"
	"os"
	"strings"

	"github.com/stevengregory/sn-update-set-extractor/internal/fileops"
	"github.com/stevengregory/sn-update-set-extractor/internal/xmlparser"
)

func main() {
	dataDir := "./data"
	outputDir := "./dist"

	files, err := os.ReadDir(dataDir)
	if err != nil {
		fmt.Println("Error reading data directory:", err)
		os.Exit(1)
	}

	for _, file := range files {
		if !strings.HasSuffix(file.Name(), ".xml") {
			continue
		}

		xmlFilePath := dataDir + "/" + file.Name()

		unload, err := xmlparser.ParseXMLFile(xmlFilePath)
		if err != nil {
			fmt.Println("Error parsing XML file", file.Name(), ":", err)
			os.Exit(1)
		}

		err = fileops.CreateDirsAndFiles(unload, outputDir)
		if err != nil {
			fmt.Println("Error creating directory structure and files:", err)
			os.Exit(1)
		}
	}

	fmt.Println("Directories and files successfully created in", outputDir)
}


internal/fileops/file_ops.go
package fileops

import (
	"encoding/xml"
	"fmt"
	"os"
	"path/filepath"

	"github.com/stevengregory/sn-update-set-extractor/internal/xmlparser"
)

func CreateDirsAndFiles(unload *xmlparser.Unload, outputDir string) error {
	widgetFileTypes := getWidgetFileTypes()

	for _, script := range unload.XMLScripts {
		parentDir := getParentDirForType(script.Type)
		if parentDir == "" {
			continue
		}

		dirPath := filepath.Join(outputDir, parentDir, script.Type)
		if err := os.MkdirAll(dirPath, 0755); err != nil {
			return err
		}

		fileExt := getFileExtForType(script.Type)
		fileName := fmt.Sprintf("%s.%s", script.TargetName, fileExt)
		filePath := filepath.Join(dirPath, fileName)

		if script.Type == "Widget" || script.Type == "Header | Footer" {
			var recordUpdate xmlparser.RecordUpdate
			err := xml.Unmarshal([]byte(script.Payload), &recordUpdate)
			if err != nil {
				fmt.Printf("Failed to parse widget: %v\n", err)
				continue
			}

			scriptName := script.TargetName

			widgetDirPath := filepath.Join(dirPath, scriptName)
			if err := os.MkdirAll(widgetDirPath, 0755); err != nil {
				return err
			}

			content := getWidgetContentType(script.Type, recordUpdate)

			for key, value := range content {
				fileName := widgetFileTypes[key]
				filePath := filepath.Join(widgetDirPath, fileName)
				if err := os.WriteFile(filePath, []byte(value), 0644); err != nil {
					return err
				}
			}
		} else if script.Type == "Fix Script" {
			var recordUpdate xmlparser.RecordUpdate
			err := xml.Unmarshal([]byte(script.Payload), &recordUpdate)
			if err != nil {
				fmt.Printf("Failed to parse fix script: %v\n", err)
				continue
			}

			jsContent := xmlparser.ExtractCDATA(recordUpdate.FixScript.Script)
			if err := os.WriteFile(filePath, []byte(jsContent), 0644); err != nil {
				return err
			}
		} else if script.Type == "UI Action" {
			var recordUpdate xmlparser.RecordUpdate
			err := xml.Unmarshal([]byte(script.Payload), &recordUpdate)
			if err != nil {
				fmt.Printf("Failed to parse UI action: %v\n", err)
				continue
			}

			jsContent := xmlparser.ExtractCDATA(recordUpdate.UIAction.Script)
			if err := os.WriteFile(filePath, []byte(jsContent), 0644); err != nil {
				return err
			}
		} else if script.Type == "Theme" {
			var recordUpdate xmlparser.RecordUpdate
			err := xml.Unmarshal([]byte(script.Payload), &recordUpdate)
			if err != nil {
				fmt.Printf("Failed to parse theme: %v\n", err)
				continue
			}

			scssContent := xmlparser.ExtractCDATA(recordUpdate.Theme.CssVariables)
			if err := os.WriteFile(filePath, []byte(scssContent), 0644); err != nil {
				return err
			}
		} else if script.Type == "Page" {
			var recordUpdate xmlparser.RecordUpdate
			err := xml.Unmarshal([]byte(script.Payload), &recordUpdate)
			if err != nil {
				fmt.Printf("Failed to parse page: %v\n", err)
				continue
			}

			scssContent := xmlparser.ExtractCDATA(recordUpdate.Page.Css)
			if err := os.WriteFile(filePath, []byte(scssContent), 0644); err != nil {
				return err
			}
		} else if script.Type == "Style Sheet" {
			var recordUpdate xmlparser.RecordUpdate
			err := xml.Unmarshal([]byte(script.Payload), &recordUpdate)
			if err != nil {
				fmt.Printf("Failed to parse style sheet: %v\n", err)
				continue
			}

			scssContent := xmlparser.ExtractCDATA(recordUpdate.StyleSheet.Css)
			if err := os.WriteFile(filePath, []byte(scssContent), 0644); err != nil {
				return err
			}
		} else {
			jsContent := xmlparser.ExtractCDATA(script.Payload)
			if err := os.WriteFile(filePath, []byte(jsContent), 0644); err != nil {
				return err
			}
		}
	}

	return nil
}

func doWidgetOperation(widgetContent xmlparser.WidgetContent) map[string]string {
	content := map[string]string{
		"client_script": xmlparser.ExtractCDATA(widgetContent.GetClientScript()),
		"css":           xmlparser.ExtractCDATA(widgetContent.GetCss()),
		"script":        xmlparser.ExtractCDATA(widgetContent.GetScript()),
		"template":      xmlparser.ExtractCDATA(widgetContent.GetTemplate()),
		"option_schema": xmlparser.ExtractCDATA(widgetContent.GetOptionSchema()),
		"link":          xmlparser.ExtractCDATA(widgetContent.GetLink()),
	}
	return content
}

func getFileExtForType(fileType string) string {
	switch fileType {
	case "Angular ng-template":
		return "html"
	case "Style Sheet":
		return "scss"
	case "Page":
		return "scss"
	case "Theme":
		return "scss"
	default:
		return "js"
	}
}

func getParentDirForType(fileType string) string {
	parentDirs := map[string]string{
		"Client Script":           "Client Development",
		"UI Script":               "Client Development",
		"Angular ng-template":     "Service Portal",
		"Widget Angular Provider": "Service Portal",
		"Header | Footer":         "Service Portal",
		"Page":                    "Service Portal",
		"Style Sheet":             "Service Portal",
		"Theme":                   "Service Portal",
		"Widget":                  "Service Portal",
		"Business Rule":           "Server Development",
		"Fix Script":              "Server Development",
		"Script Include":          "Server Development",
		"UI Action":               "Server Development",
		"Scripted REST Resource":  "Inbound Integrations",
	}
	return parentDirs[fileType]
}

func getWidgetContentType(fileType string, recordUpdate xmlparser.RecordUpdate) map[string]string {
	var content map[string]string
	switch fileType {
	case "Widget":
		content = doWidgetOperation(recordUpdate.Widget)
	case "Header | Footer":
		content = doWidgetOperation(recordUpdate.HeaderFooter)
	}
	return content
}

func getWidgetFileTypes() map[string]string {
	return map[string]string{
		"client_script": "client_script.js",
		"css":           "style.scss",
		"script":        "server_script.js",
		"template":      "template.html",
		"option_schema": "option_schema.json",
		"link":          "link.js",
	}
}
